stan/math/prim/scal/meta/contains_nonconstant_struct.hpp

```diff
@@ -2,18 +2,16 @@
 #define STAN_MATH_PRIM_SCAL_META_CONTAINS_NONCONSTANT_STRUCT_HPP
 
 #include <stan/math/prim/scal/meta/is_constant_struct.hpp>
-
+#include <stan/math/prim/scal/meta/or.hpp>
+#include <type_traits>
 namespace stan {
-
-template <typename T1, typename T2 = double, typename T3 = double,
-          typename T4 = double, typename T5 = double, typename T6 = double>
-struct contains_nonconstant_struct {
-  enum {
-    value = !is_constant_struct<T1>::value || !is_constant_struct<T2>::value
-            || !is_constant_struct<T3>::value || !is_constant_struct<T4>::value
-            || !is_constant_struct<T5>::value || !is_constant_struct<T6>::value
-  };
-};
+/**
+ * Metaprogram to determine if any of the
+ * provided types have a base scalar
+ * type that can not be assigned to type double.
+ */
+template <typename... T>
+using contains_nonconstant_struct = math::or_not_<is_constant_struct<T>...>;
```

Comment:
[optional]
Rather than using `or_not` (which again shouldn't have trailing underscore), how about using `is_nonconstant_struct` and just conjoining?